Sources/NIOCore/FileHandle.swift

```diff
@@ -54,6 +57,7 @@ public final class NIOFileHandle: FileDescriptor {
         assert(!self.isOpen, "leaked open NIOFileHandle(descriptor: \(self.descriptor)). Call `close()` to close or `takeDescriptorOwnership()` to take ownership and close by some other means.")
     }
 
+#if !os(WASI)
```

Comment:
Same note here: generally it's preferable to have APIs not simply be absent, but to misbehave when not present. In this case, `duplicate` will presumably _become_ available, and it'd be nicer if this just threw instead. The only time that isn't my preference is when an API literally cannot be spelled, e.g. the return or argument types are missing.

Relatedly, it's best if we put the `#if os` at the _lowest_ possible level. In this case, we should arrange for `dup` to throw or error out, and let the rest of the code path handle that appropriately, which minimises the OS customisation we have to do.

Reply:
Ditto, what's the benefit of preferring obscured runtime errors when we can easily make these compile-time errors and make unavailability obvious to a library user?